File: code/eprogramoflifeapp/app/src/main/java/domain/WeeklyCommitment.java
package domain;

import java.util.Calendar;
import java.util.Date;

/**
 * Created by mmfsf on 3/29/17.
 */

public class WeeklyCommitment extends Commitment {
    public WeeklyCommitment(String name) {
        super(name);
        this.frequency = Frequency.Weekly;
    }

    public Level getPerformed(Date date)
    {
        String key = GetKey(date);

        Level level = this.performed.get(key);
        if(level == null)
        {
            level = Level.NotDone;
        }
        return level;
    }

    public void Point(Date date, Level level)
    {
        String key = GetKey(date);
        this.performed.put(key, level);
    }

    private String GetKey(Date date)
    {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        int year = calendar.get(Calendar.YEAR);
        int numberOfDays = calendar.get(Calendar.MONTH) * 30;
        int week = (int)Math.ceil(numberOfDays / 7.0);

        return String.format("%d%d", year, week);
    }
}

File: code/eprogramoflifeweb/epl.core/Domain/WeeklyCommitment.cs
using epl.core.ValuesObjects;
using System;
using System.Collections.Generic;
using System.Globalization;

namespace epl.core.Domain
{
  public class WeeklyCommitment : Commitment
  {
    protected override string KeyFormat { get => "yyyy"; }

    public WeeklyCommitment(string name) : base(name)
    {
      this.Frequency = Frequency.Weekly;
    }

    public WeeklyCommitment(string name, string description) : base(name, description)
    {
      this.Frequency = Frequency.Weekly;
    }

    public override Level GetPoint(DateTime date)
    {
      var week = CultureInfo.InvariantCulture.Calendar.GetWeekOfYear(date, CalendarWeekRule.FirstDay, DayOfWeek.Sunday);
      var key = $"{date.ToString(this.KeyFormat)}-{week}";
      return this.Performed.ContainsKey(key) ? this.Performed[key] : Level.NotDone;
    }

    public override void Point(DateTime date, Level level)
    {
      var week = CultureInfo.InvariantCulture.Calendar.GetWeekOfYear(date, CalendarWeekRule.FirstDay, DayOfWeek.Sunday);
      var key = $"{date.ToString(this.KeyFormat)}-{week}";
      
      if (this.Performed.ContainsKey(key))
      {
        this.Performed[key] = level;
      }
      else
      {
        this.Performed.Add(key, level);
      }
    }

  }
}
